pkg/controller/networkpolicy/adminnetworkpolicy.go

```diff
@@ -0,0 +1,405 @@
+	defer n.heartbeat("addBANP")
+	banp := obj.(*v1alpha1.BaselineAdminNetworkPolicy)
+	klog.InfoS("Processing BaselineAdminNetworkPolicy ADD event", "banp", banp.Name)
+	n.enqueueInternalNetworkPolicy(getBANPReference(banp))
+}
+
+// updateBANP receives BaselineAdminNetworkPolicy UPDATE events and enqueues a
+// reference of the BaselineAdminNetworkPolicy to trigger its process.
+func (n *NetworkPolicyController) updateBANP(_, cur interface{}) {
+	defer n.heartbeat("updateBANP")
+	curBANP := cur.(*v1alpha1.BaselineAdminNetworkPolicy)
+	klog.InfoS("Processing AdminNetworkPolicy UPDATE event", "banp", curBANP.Name)
+	n.enqueueInternalNetworkPolicy(getBANPReference(curBANP))
+}
+
+// deleteBANP receives BaselineAdminNetworkPolicy DELETE events and enqueues a
+// reference of the BaselineAdminNetworkPolicy to trigger its process.
+func (n *NetworkPolicyController) deleteBANP(old interface{}) {
+	banp, ok := old.(*v1alpha1.BaselineAdminNetworkPolicy)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting BaselineAdminNetworkPolicy, invalid type: %v", old)
+			return
+		}
+		banp, ok = tombstone.Obj.(*v1alpha1.BaselineAdminNetworkPolicy)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting BaselineAdminNetworkPolicy, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
+	defer n.heartbeat("deleteBANP")
+	klog.InfoS("Processing BaselineAdminNetworkPolicy DELETE event", "banp", banp.Name)
+	n.enqueueInternalNetworkPolicy(getBANPReference(banp))
+}
+
+// anpHasNamespaceLabelRule returns whether an AdminNetworkPolicy has rules defined by
+// advanced Namespace selection (sameLabels and notSameLabels)
+func anpHasNamespaceLabelRule(anp *v1alpha1.AdminNetworkPolicy) bool {
+	for _, ingress := range anp.Spec.Ingress {
+		for _, peer := range ingress.From {
+			if peer.Namespaces != nil && (len(peer.Namespaces.SameLabels) > 0 || len(peer.Namespaces.NotSameLabels) > 0) {
+				return true
+			}
+		}
+	}
+	for _, egress := range anp.Spec.Egress {
+		for _, peer := range egress.To {
+			if peer.Namespaces != nil && (len(peer.Namespaces.SameLabels) > 0 || len(peer.Namespaces.NotSameLabels) > 0) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// banpHasNamespaceLabelRule returns whether a BaselineAdminNetworkPolicy has rules defined by
+// advanced Namespace selection (sameLabels and notSameLabels)
+func banpHasNamespaceLabelRule(banp *v1alpha1.BaselineAdminNetworkPolicy) bool {
+	for _, ingress := range banp.Spec.Ingress {
+		for _, peer := range ingress.From {
+			if peer.Namespaces != nil && (len(peer.Namespaces.SameLabels) > 0 || len(peer.Namespaces.NotSameLabels) > 0) {
+				return true
+			}
+		}
+	}
+	for _, egress := range banp.Spec.Egress {
+		for _, peer := range egress.To {
+			if peer.Namespaces != nil && (len(peer.Namespaces.SameLabels) > 0 || len(peer.Namespaces.NotSameLabels) > 0) {
+				return true
+			}
+		}
+	}
+	return false
+}
```

Comment:
there are lots of duplicated codes, you may wrap a method with a `[]Rules` parameter and check if they has advanced Namespace selection.

Reply:
There are tons of duplications, however most are necessary given that the `Rules` you refer to are actually of different go types, namely `AdminNetworkPolicyIngressRule`, `AdminNetworkPolicyEgressRule`, `BaselineAdminNetworkPolicyIngressRule` and `BaselineAdminNetworkPolicyEgressRule`. I could define a interface and have each type implement functions like `getNamespaces`, but feels like an overkill to me for this specific purpose